### PR TITLE
Dbaas 2852

### PR DESCRIPTION
**Pull request by sharmanirek**

### CODE REVIEW

**Review by abaveja313 — APPROVED** [review]

File: splicemachine/ml/stats.py [hunk 1]
```diff
@@ -132,6 +132,72 @@ def _transform(self, dataset):
 
         return dataset
 
+# This is largely useless unless I time it and it turns out to be faster
+class myIndexToString(Transformer, HasInputCol, HasOutputCol):
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]: If we keep this in it needs docstrings with an example use in the docstring

File: splicemachine/ml/stats.py [hunk 2]
```diff
@@ -132,6 +132,72 @@ def _transform(self, dataset):
 
         return dataset
 
+# This is largely useless unless I time it and it turns out to be faster
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]: remove comment

File: splicemachine/ml/stats.py [hunk 3]
```diff
@@ -132,6 +132,72 @@ def _transform(self, dataset):
...
+class myIndexToString(Transformer, HasInputCol, HasOutputCol):
+    @keyword_only
+    def __init__(self, predCol=None, labelCol=None, stringCol = None, reconstructed_base = '_string'):
+        super(myIndexToString, self).__init__()
+        # kwargs = self._input_kwargs
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]: remove comments

File: splicemachine/ml/stats.py [hunk 4]
```diff
@@ -132,6 +132,72 @@ def _transform(self, dataset):
...
+        stringCol = self.stringCol
+        reconstructed_base = self.reconstructed_base
+
+        joinedout = dataset.join(dataset.select([labelCol, stringCol])\
+                            .withColumnRenamed(label,predCol)\
+                            .withColumnRenamed(stringCol, predCol + reconstructed_base), predCol)\
+                            .dropDuplicates()
+
+        return joinedout
+# Now I'll make oversampling a transformer to include it in the pipeline
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]: remove comment

File: splicemachine/ml/stats.py (hunk 1 above)
Comment (Ben-Epstein) [reply splicemachine/ml/stats.py]: Include params for __init__ function

File: splicemachine/ml/stats.py [hunk 5]
```diff
@@ -132,6 +132,72 @@ def _transform(self, dataset):
...
+
+    Follows: https://spark.apache.org/docs/latest/ml-pipeline.html#transformers
+
+    :param Transformer: Inherited Class
+    :param HasInputCol: Inherited Class
+    :param HasOutputCol: Inherited Class
+    :return: Transformed PySpark Dataframe With Original String Indexed Variables
+    """
+    @keyword_only
+    def __init__(self, label=None, ratio = None, majorityLabel = None, minorityLabel = None, withReplacement = None):
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]: add docstring with param description

File: splicemachine/ml/utilities.py [hunk 6]
```diff
@@ -175,7 +175,7 @@ def plotROC(self, trainingSummary, ax):
         """
         Plots the receiver operating characteristic curve for the trained classifier
 
-        :param trainingSummary: TrainingSummary object accessed by .summary after fitting a binary classification object
+        :param trainingSummary: TrainingSummary object accessed by .summary after fitting a binary classification object ONLY VALID FOR LOGISTIC REGRESSION
```
Comment (Ben-Epstein) [inline splicemachine/ml/utilities.py]: Change this function to take in a model, and check that the model is a logistic regression -- throw an assertion error if it's not. If it is, get the summary within the function

**Review by Ben-Epstein — CHANGES_REQUESTED** [review]
See comments

**Comment by sharmanirek** [comment]
Addressed Comments

File: splicemachine/ml/stats.py [hunk 7]
```diff
@@ -4,21 +4,129 @@
...
+    >>> rounder = Rounder(predictionCol = "prediction", labelCol = "label", clipPreds = True)
+    >>> rounder.transform(dataset).show()
+    +----------+-----+
+    |prediction|label|
+    +----------+-----+
+    |       0.0|  0.0|
+    |       1.0|  1.0|
+    |       2.0|  2.0|
+    |       1.0|  0.0|
+    |       2.0|  0.0|
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py:67]: should this prediction be a 3.0 here?

File: splicemachine/ml/stats.py [hunk 8]
```diff
@@ -132,6 +240,262 @@ def _transform(self, dataset):
...
+
+        return [singleTask] * len(epm)
+
+    def _fit(self, dataset):
+        """Performs k-fold crossvaldidation on simple oversampled dataset
+
+        :param dataset: full dataset
+        :return: CrossValidatorModel containing the fitted BestModel with the average of the primary and alternate metrics in a list of tuples in the format: [(paramComb1_average_primary_metric, [paramComb1_average_altmetric1,paramComb1_average_altmetric2]), (paramComb2_average_primary_metric, [paramComb2_average_altmetric1,paramComb2_average_altmetric2])]
+        """
+        est = self.getOrDefault(self.estimator)
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py:439]: You're assuming the estimator and estimatorParamMaps are available, so they shouldn't default to None in the function definition. They should be required

File: splicemachine/ml/stats.py [hunk 9]
```diff
@@ -132,6 +240,262 @@ def _transform(self, dataset):
...
+    >>> AUCevaluator = BinaryClassificationEvaluator(metricName = 'areaUnderROC')
+    >>> ACCevaluator = MulticlassClassificationEvaluator(metricName="accuracy")
+    >>> cv = OverSampleCrossValidator(estimator=lr, estimatorParamMaps=grid, evaluator=AUCevaluator, altEvaluators = [PRevaluator, ACCevaluator],parallelism=2,seed = 1234)
+    >>> cvModel = cv.fit(dataset)
+    >>> print(cvModel.avgMetrics)
+    [(0.5, [0.5888888888888888, 0.3888888888888889]), (0.806878306878307, [0.8556863149300125, 0.7055555555555556])]
+    >>> print(AUCevaluator.evaluate(cvModel.transform(dataset)))
+    0.8333333333333333
+    """
+    def __init__(self, estimator=None, estimatorParamMaps=None, evaluator=None, numFolds=3, seed=None, parallelism=1, collectSubModels=False, labelCol = 'label', altEvaluators = None, overSample = True):
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]: These shouldn't all be None/have default values. Whichever must be provided for the function to operate shouldn't have default values.

File: splicemachine/ml/stats.py [hunk 10]
```diff
@@ -132,6 +240,262 @@ def _transform(self, dataset):
...
+        """ Initialize Self
+
+        :param estimator: Machine Learning Model, defaults to None
+        :param estimatorParamMaps: paramMap to search, defaults to None
+        :param evaluator: primary model evaluation metric, defaults to None
+        :param numFolds: number of folds to perform, defaults to 3
+        :param seed: random state, defaults to None
+        :param parallelism: number of threads, defaults to 1
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py:387]: Should parallelism default to the same as numFolds (3)?

**Review by Ben-Epstein — CHANGES_REQUESTED** [review]

File: splicemachine/ml/stats.py (hunk 7 above)
Comment (sharmanirek) [reply splicemachine/ml/stats.py:67]: Param clipPreds= True forces the max prediction to be no greater than the max label (either specified by param maxLabel, or (default) max(F.col(label))

Comment (sharmanirek) [reply splicemachine/ml/stats.py:67]: Param clipPreds= True forces the max prediction to be no greater than the max label (either specified by param maxLabel, or (default) max(F.col(label))

File: splicemachine/ml/stats.py (hunk 10 above)
Comment (sharmanirek) [reply splicemachine/ml/stats.py:387]: That is a good point

File: splicemachine/ml/stats.py (hunk 8 above)
Comment (sharmanirek) [reply splicemachine/ml/stats.py:439]: Good point. Thanks

Comment (sharmanirek) [reply splicemachine/ml/stats.py:439]: Good point. Thanks

File: splicemachine/ml/stats.py (hunk 9 above)
Comment (sharmanirek) [reply splicemachine/ml/stats.py]: Good point

Comment (sharmanirek) [reply splicemachine/ml/stats.py]: Good point

Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]:
```suggestion
    def __init__(self, estimator, estimatorParamMaps, evaluator, numFolds=3, seed=None, parallelism=1, collectSubModels=False, labelCol = 'label', altEvaluators = None, overSample = True):
```

**Review by Ben-Epstein — CHANGES_REQUESTED** [review]

File: splicemachine/ml/stats.py [hunk 11]
```diff
@@ -132,6 +240,262 @@ def _transform(self, dataset):
...
+    >>> AUCevaluator = BinaryClassificationEvaluator(metricName = 'areaUnderROC')
+    >>> ACCevaluator = MulticlassClassificationEvaluator(metricName="accuracy")
+    >>> cv = OverSampleCrossValidator(estimator=lr, estimatorParamMaps=grid, evaluator=AUCevaluator, altEvaluators = [PRevaluator, ACCevaluator],parallelism=2,seed = 1234)
+    >>> cvModel = cv.fit(dataset)
+    >>> print(cvModel.avgMetrics)
+    [(0.5, [0.5888888888888888, 0.3888888888888889]), (0.806878306878307, [0.8556863149300125, 0.7055555555555556])]
+    >>> print(AUCevaluator.evaluate(cvModel.transform(dataset)))
+    0.8333333333333333
+    """
+    def __init__(self, estimator, estimatorParamMaps, evaluator, numFolds=3, seed=None, parallelism=1, collectSubModels=False, labelCol = 'label', altEvaluators = None, overSample = True):
```
Comment (Ben-Epstein) [inline splicemachine/ml/stats.py]:
```suggestion
    def __init__(self, estimator, estimatorParamMaps, evaluator, numFolds=3, seed=None, parallelism=3, collectSubModels=False, labelCol = 'label', altEvaluators = None, overSample = True):
```

**Review by Ben-Epstein — APPROVED** [review]

**Review by Ben-Epstein — APPROVED** [review]

**Pull request merged**